### PR TITLE
[G-API] Handle errors in IEBackend & modeling tool

### PR DESCRIPTION
**Pull request by TolyaTalamanov**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ ] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### Build configuration
force_builders=Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

Xbuild_image:Custom=centos:7
Xbuildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

build_image:Custom=ubuntu-openvino-2021.4.1:20.04
build_image:Custom Win=openvino-2021.4.1
build_image:Custom Mac=openvino-2021.4.1

test_modules:Custom=gapi,python2,python3,java
test_modules:Custom Win=gapi,python2,python3,java
test_modules:Custom Mac=gapi,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*

### CODE REVIEW

File: modules/gapi/src/backends/ie/giebackend.cpp [hunk 1]
```diff
@@ -216,6 +216,39 @@ inline void copyFromIE(const IE::Blob::Ptr &blob, MatType &mat) {
     }
 }
 
+template <typename MapT>
```
Comment (mpashchenkov) [inline modules/gapi/src/backends/ie/giebackend.cpp:219]: MB need to add check for input types here. But it is backend and you won't try to send not map type. Up to you.

**Review by mpashchenkov — COMMENTED** [review]
Minor

File: modules/gapi/src/backends/ie/giebackend.cpp (hunk 1 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/ie/giebackend.cpp:219]: Yep, probably it's needed to add SFINAE in order to check that `MatT` is equal to `ConstInputsDataMap`, `ConstOutputsDataMap`, `InputsDataMap`, `OutputsDataMap` and check that `layer_type` is either `input` or `output`, but I just didn't overcomplicate the code, this function isn't supposed to be reusable for anything else

File: modules/gapi/src/backends/ie/giebackend.cpp [hunk 2]
```diff
@@ -293,6 +326,16 @@ struct IEUnit {
                          params.num_in &&
                         "Number of layers to reshape must be less than or equal to number of inputs");
         }
+
+        if (params.kind == cv::gapi::ie::detail::ParamDesc::Kind::Load) {
+            checkInputLayerNames(net.getInputsInfo() , params.input_names);
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/ie/giebackend.cpp]: Done

**Review by mpashchenkov — APPROVED** [review]

**Review by dmatveev — APPROVED** [review]
👍

File: modules/gapi/src/backends/ie/giebackend.cpp [hunk 3]
```diff
@@ -652,14 +698,22 @@ void cv::gimpl::ie::RequestPool::execute(cv::gimpl::ie::RequestPool::Task&& t) {
...
     t.run(request);
 }
 
 void cv::gimpl::ie::RequestPool::callback(cv::gimpl::ie::RequestPool::Task task,
-                                          InferenceEngine::InferRequest& request,
-                                          size_t id) {
+                                          size_t id,
+                                          IE::InferRequest request,
+                                          IE::StatusCode /* code */) {
+    // FIXME: Ignore IE::StatusCode so far,
```
Comment (TolyaTalamanov) [inline modules/gapi/src/backends/ie/giebackend.cpp:714]: @dmatveev Does it make sense to put some log here in order to notify user about problem?

File: modules/gapi/src/backends/ie/giebackend.cpp [hunk 4]
```diff
@@ -652,21 +698,38 @@ void cv::gimpl::ie::RequestPool::execute(cv::gimpl::ie::RequestPool::Task&& t) {
...
+    using callback_t = std::function<void(IE::InferRequest, IE::StatusCode)>;
     request.SetCompletionCallback(
-            std::bind(&cv::gimpl::ie::RequestPool::callback, this, t, std::ref(request), id));
+            static_cast<callback_t>(
+                std::bind(&cv::gimpl::ie::RequestPool::callback, this,
+                          t, id, _1, _2)));
     t.run(request);
 }
 
 void cv::gimpl::ie::RequestPool::callback(cv::gimpl::ie::RequestPool::Task task,
```
Comment (TolyaTalamanov) [inline modules/gapi/src/backends/ie/giebackend.cpp:710]: Filed a task to fix this mess

**Comment by TolyaTalamanov** [comment]
@alalek Could you merge it?

File: modules/gapi/samples/pipeline_modeling_tool.cpp [hunk 5]
```diff
@@ -393,6 +398,16 @@ int main(int argc, char* argv[]) {
 
         for (size_t i = 0; i < threads.size(); ++i) {
             threads[i].join();
+        }
+
+        for (size_t i = 0; i < threads.size(); ++i) {
+            if (eptrs[i] != nullptr) {
+                try {
+                    std::rethrow_exception(eptrs[i]);
+                } catch (std::exception& e) {
```
Comment (alalek) [inline modules/gapi/samples/pipeline_modeling_tool.cpp:407]: const reference in catch

File: modules/gapi/src/backends/ie/giebackend.cpp [hunk 6]
```diff
@@ -652,21 +698,38 @@ void cv::gimpl::ie::RequestPool::execute(cv::gimpl::ie::RequestPool::Task&& t) {
...
+            throw std::logic_error("IE::InferRequest finished with not OK status");
+        }
+        task.callback(request);
+        // NB: IE::InferRequest keeps the callback until the new one is set.
+        // Since user's callback might keep resources that should be released,
+        // need to destroy its after execution.
+        // Let's set the empty one to cause the destruction of a callback.
+        request.SetCompletionCallback([](){});
+        m_idle_ids.push(id);
+    } catch (std::exception& e) {
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/ie/giebackend.cpp]: Done

**Comment by TolyaTalamanov** [comment]
@alalek CI is green

**Pull request merged**